### PR TITLE
Implementation of Basic04

### PR DESCRIPTION
**Pull request by vlevigneron**
Fixes zonemaster/zonemaster-engine#763

Need to be done:
=> Unit tests
=> translation of new messages

### CODE REVIEW

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 1]
```diff
@@ -128,57 +153,85 @@ sub metadata {
 } ## end sub metadata
 
 Readonly my %TAG_DESCRIPTIONS => (
+    A_QUERY_NO_RESPONSES => sub {
+        __x    # BASIC:A_QUERY_NO_RESPONSES
+          'Nameservers did not respond to A query.';
+    },
+    B04_MISSING_NS_RECORD => sub {
+        __x    # BASIC:B04_MISSING_NS_RECORD
+          'Nameserver {ns} repond to a NS query with no records in answer.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:162]: What data type is `ns`? According to [logentry_args] `ns` should be "domain name and IP address pair", but I do not get that impression from the code below were it is used.

```
The name and IP address of a name server, separated by "/".
```






[logentry_args]: https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 2]
```diff
@@ -511,6 +766,11 @@ Checks if at least one of the nameservers pointed out by the parent zone gives a
 tested zone (that is, if we're testing C<example.org> this test will as for A records for C<www.example.org>). This test is only run if the
 L<basic02> test has I<failed>.
 
+=item basic04
+
+Query all nameservers pointed out by the parent zone or found in delegation for NS and/or SOA records. Initially done in several
+test cases, these tests should be done only here.
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:772]: "Initially" --> "previously"?

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 3]
```diff
@@ -55,6 +55,9 @@ sub all {
                 }
               );
         }
+
+        push @results, $class->basic04( $zone );
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:59]: This should before basic02 according to the specification.

**Review by matsduf — CHANGES_REQUESTED** [review]

File: lib/Zonemaster/Engine/Test/Basic.pm (hunk 1 above)
Comment (matsduf) [reply lib/Zonemaster/Engine/Test/Basic.pm:162]: 'Nameserver {ns} reponds to an NS query with no NS records in the answer section.', @_;

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 4]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+        __x    # BASIC:A_QUERY_NO_RESPONSES
+          'Nameservers did not respond to A query.';
+    },
+    B04_MISSING_NS_RECORD => sub {
+        __x    # BASIC:B04_MISSING_NS_RECORD
+          'Nameserver {ns} repond to a NS query with no records in answer.', @_;
+    },
+    B04_MISSING_SOA_RECORD => sub {
+        __x    # BASIC:B04_MISSING_SOA_RECORD
+          'Nameserver {ns} repond to a SOA query with no records in answer.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:166]: 'Nameserver {ns} reponds to an SOA query with no SOA record in answer section.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 5]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_NO_RESPONSE => sub {
+        __x    # BASIC:B04_NO_RESPONSE
+          'Nameserver {ns} did not respond over UDP and TCP.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:170]: 'Nameserver {ns} does not respond over neither UDP nor TCP.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 6]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_NO_RESPONSE_NS_QUERY => sub {
+        __x    # BASIC:B04_NO_RESPONSE_NS_QUERY
+          'No response from nameserver {ns} on NS queries.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:174]: 'Nameserver {ns} does not respond to NS queries.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 7]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_NO_RESPONSE_SOA_QUERY => sub {
+        __x    # BASIC:B04_NO_RESPONSE_SOA_QUERY
+          'No response from nameserver {ns} on SOA queries.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:178]: 'Nameserver {ns} does not respond to SOA queries.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 8]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_NS_RECORD_NOT_AA => sub {
+        __x    # BASIC:B04_NS_RECORD_NOT_AA
+          'Nameserver {ns} response is not authoritative on NS query.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:182]: 'Nameserver {ns} does not give an authoritative response on an NS query.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 9]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_RESPONSE_TCP_NOT_UDP => sub {
+        __x    # BASIC:B04_RESPONSE_TCP_NOT_UDP
+          'Nameserver {ns} did not respond over UDP.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:186]: 'Nameserver {ns} does not respond over UDP.'

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 10]
```diff
@@ -128,57 +153,85 @@ sub metadata {
...
+    },
+    B04_SOA_RECORD_NOT_AA => sub {
+        __x    # BASIC:B04_SOA_RECORD_NOT_AA
+          'Nameserver {ns} response is not authoritative on SOA query.', @_;
+    },
+    B04_UNEXPECTED_RCODE_NS_QUERY => sub {
+        __x    # BASIC:B04_UNEXPECTED_RCODE_NS_QUERY
+          'Nameserver {ns} responds with an unexpected RCODE ({rcode}) on a NS query.', @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:194]: "an NS query"

File: lib/Zonemaster/Engine/Test/Basic.pm (hunk 1 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Basic.pm:162]: @matsduf I just double checked and 'ns' is as expected (The name and IP address of a name server, separated by "/").

File: lib/Zonemaster/Engine/Test/Basic.pm (hunk 2 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Basic.pm:772]: Doone

File: lib/Zonemaster/Engine/Test/Basic.pm (hunk 3 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/Basic.pm:59]: Fixed in next commit

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
## Release testing for v2021.1

Works as expected.

**Pull request merged**